### ISSUE TITLE
clang++-19 -std=gnu++14 -Wc++20-extensions wrongly warns about GNU variadic macro extension

### ISSUE THREAD

**Issue by jinkama-san** [llvm/llvm-project#109677]
Hi,

for simple code like

```
#ifdef DEBUG
# define DBG(fmt, ...) printf(fmt, ##__VA_ARGS__)
#else
# define DBG(fmt, ...)
#endif
#include <stdio.h>
int main()
{
        DBG("Test\n");
}
```

compiled with `clang++-19 -std=gnu++14  test-variadic-macro.cpp -Wc++20-extensions` now this incorrect (for C++14 *with* GNU extensions) warning shows:

```
test-variadic-macro.cpp:11:14: warning: passing no argument for the '...' parameter of a variadic macro is a C++20 extension [-Wc++20-extensions]
   11 |         DBG("Test\n");
      |                     ^
test-variadic-macro.cpp:4:10: note: macro 'DBG' defined here
    4 | # define DBG(fmt, ...)
      |          ^
1 warning generated.
```

This breaks a lot of code in release mode compiles. Also adding  `-Wno-gnu-zero-variadic-macro-arguments` after `-Wc++20-extensions` doesn't help here. All earlier clang versions I tested including clang-18.1.8 got this right and didn't warn.
Maybe related to issue #76375

**Comment by shafik** [comment]
Look like @Sirraide added this in 5231005193afb

The warning looks correct but maybe we need a way to turn it off.


CC @erichkeane

**Comment by erichkeane** [comment]
>> Also adding -Wno-gnu-zero-variadic-macro-arguments after -Wc++20-extensions doesn't help here.

This is likely the bug here, the warning is correct.  Its a GNU extension but it is also a C++20 extension.  But the fact that we're not able to disable it with its flag if the higher-level flag is enabled is problematic.

**Comment by Sirraide** [comment]
> Look like @Sirraide added this in [5231005](https://github.com/llvm/llvm-project/commit/5231005193afb8db01afe9a8a1aa308d25f60ba1)

Iirc the warning was already there before, and I just changed it to allow it for C23 or sth like that, but that was a while ago, so I’m not 100% sure anymore.

**Comment by jinkama-san** [comment]
Or maybe, regardless of any warning options, suppress the warning always if none of the named arguments are actually used?
For macros there is never (AFAICT) a warning when a named argument is not used, so a warning after an unused named argument doesn't make much sense to me.

**Comment by toughengineer** [comment]
I accidentally discovered that **Clang 20** _does suppress_ the warning/error with `-Wno-gnu-zero-variadic-macro-arguments`.
Just so you know.

See
https://godbolt.org/z/Wazoe63q1
